mhs-reference-implementation/mhs/parser/ebxml_message_parser.py

```diff
@@ -1,38 +1,137 @@
+        self._extract_ebxml_value_is_present(xml_tree, "AckRequested", extracted_values, ACK_REQUESTED)
+        self._extract_attribute(xml_tree, "AckRequested", SOAP_NAMESPACE, "actor", extracted_values, ACK_SOAP_ACTOR)
 
-class EbXmlRequestMessageParser(EbXmlMessageParser):
-    """A component that extracts information from ebXML acknowledgement messages."""
+        return extracted_values
+
+    def _path_to_ebxml_element(self, name, parent=None):
+        path = ".//"
+
+        if parent is not None:
+            path += EBXML_NAMESPACE + ":" + parent + "/"
+
+        path += EBXML_NAMESPACE + ":" + name
+
+        return path
+
+    def _extract_ebxml_text_value(self, xml_tree, element_name, values_dict, key, parent=None):
+        xpath = self._path_to_ebxml_element(element_name, parent=parent)
+        text_value = xml_tree.findtext(xpath, namespaces=NAMESPACES)
+        if text_value is not None:
+            values_dict[key] = text_value
+
+    def _extract_ebxml_value_is_present(self, xml_tree, element_name, values_dict, key):
+        xpath = self._path_to_ebxml_element(element_name)
+        value = xml_tree.find(xpath, NAMESPACES)
+        if value is not None:
+            values_dict[key] = True
+
+    def _extract_attribute(self, xml_tree, element_name, attribute_namespace, attribute_name, values_dict, key):
+        xpath = self._path_to_ebxml_element(element_name)
+        element = xml_tree.find(xpath, NAMESPACES)
+        if element is not None:
+            values_dict[key] = element.attrib["{" + NAMESPACES[attribute_namespace] + "}" + attribute_name]
 
-    # TODO: Will contain the path the the ack-specific schematron/XSLT file.
 
-    def parse_message(self, message):
+class EbXmlRequestMessageParser(EbXmlMessageParser):
+    """A component that extracts information from ebXML request messages."""
+
+    def parse_message(self, headers, message):
         """Parse the provided ebXML request message and extract a dictionary of values from it.
 
+        :param headers A dictionary of headers received with the message.
         :param message: The message to be parsed.
         :return: A dictionary of values extracted from the message.
         """
-        return {
-            FROM_PARTY_ID: "YES-0000806",
-            CPA_ID: "S1001A1630",
-            CONVERSATION_ID: "10F5A436-1913-43F0-9F18-95EA0E43E61A",
-            MESSAGE_ID: "C614484E-4B10-499A-9ACD-5D645CFACF61",
-            TIMESTAMP: "2019-05-04T20:55:16Z"
-        }
+        msg = self._parse_mime_message(headers, message)
+        ebxml_part, payload_part = self._extract_message_parts(msg)
+        extracted_values = super().parse_message(headers, ebxml_part)
+
+        if payload_part:
+            extracted_values[MESSAGE] = payload_part
+
+        logging.debug("Extracted values from message: %s", extracted_values)
+        return extracted_values
+
+    def _parse_mime_message(self, headers, message):
```

Comment:
Perhaps a docstring to explain what this function will be returning.

Reply:
Yep, fair point. I'll add one.

Reply:
Added